harness/api_checkers.go

```diff
@@ -0,0 +1,437 @@
+// Package harness provides a suite of API compatibility checks. They were originally
+// ported from the OpenTracing Python library's "harness" module.
+package harness
+
+import (
+	"bytes"
+	"fmt"
+	"testing"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+// APICheckCapabilities describes options used by APICheckSuite when testing a Tracer.
+type APICheckCapabilities struct {
+	CheckBaggageValues bool          // whether to check for propagation of baggage values
+	CheckExtract       bool          // whether to check if extracting contexts from carriers works
+	CheckInject        bool          // whether to check if injecting contexts works
+	Probe              APICheckProbe // optional interface providing methods to check recorded data
+}
+
+// APICheckProbe exposes methods for testing data recorded by a Tracer.
+type APICheckProbe interface {
+	// SameTrace helps tests assert that this tracer's spans are from the same trace.
+	SameTrace(first, second opentracing.Span) bool
+	// SameSpanContext helps tests assert that a span and a context are from the same trace and span.
+	SameSpanContext(opentracing.Span, opentracing.SpanContext) bool
+}
+
+// APICheckSuite is a testify suite for checking a Tracer against the OpenTracing API.
+type APICheckSuite struct {
+	suite.Suite
+	opts      APICheckCapabilities
+	newTracer func() (tracer opentracing.Tracer, closer func())
+	tracer    opentracing.Tracer
+	closer    func()
+}
+
+// NewAPICheckSuite returns a testify suite for checking a Tracer against the OpenTracing API.
+// It is provided a function that will be executed to create and destroy a tracer for each test
+// in the suite, and API test options described by APICheckCapabilities.
+func NewAPICheckSuite(
+	newTracer func() (tracer opentracing.Tracer, closer func()),
+	opts ...APICheckOption,
+) *APICheckSuite {
+	s := &APICheckSuite{newTracer: newTracer}
+	for _, o := range opts {
+		o.Apply(s)
+	}
+	return s
+}
+
+// APICheckOption instances may be passed to NewAPICheckSuite.
+type APICheckOption interface {
+	Apply(*APICheckSuite)
```

Comment:
the "options factory" is just a namespacing trick, you don't need it if the namespace is small enough like that of `harness`. I was simply referring to defining Option as a function, rather than interface, it's more Go-ish (I guess we dropped the ball when doing `StartSpanOption` :-) ). Not a big deal either way.